protoquant adding dynamically_quantize_per_channel

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #24
* #22
* #21
* __->__ #20

Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

HDCharles committed:
protoquant adding dynamically_quantize_per_channel

Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function.

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function.

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function.

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on ao_experimental/quant_primitives.py:
```diff
@@ -56,6 +56,78 @@ def dynamically_quantize_per_tensor(
+        maxs = x2.max(dim=1).values
+        return mins, maxs
+
+    min_val, max_val = get_min_max_per_channel(x, axis=axis)
```
just curious, what's the context for using this function?  Can it be done simpler by just doing something like `min(dim=axis)`?

HDCharles replied:
no the dim=axis argument only allows you to remove one dimension at a time when you call max, so if you had 4 dimensional input, one dimension of which was channel, you'd need to call that 3 times without this function. Not sure how important that kind of support is though.

vkuzo replied:
does https://pytorch.org/docs/stable/generated/torch.amax.html solve this? it supports reducing over multiple dimensions

HDCharles replied:
yeah that works, fixed

vkuzo commented on ao_experimental/quant_primitives.py:
```diff
@@ -56,6 +56,78 @@ def dynamically_quantize_per_tensor(
+    # reference: https://github.com/pytorch/pytorch/blob/a3989b2802a5b32d8793557ddb5aba36298ef2be/torch/ao/quantization/observer.py#L330
+    max_val_pos = torch.max(max_val, -min_val)
+
+    scales = (
```
nit: add comment for precision handling

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757989).

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757989).

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757989).

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757989).

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757989).

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757989).

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757989).

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding dynamically_quantize_per_channel"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function generalizes that function and fixes some
numerical issues. Specifically it allows for arbitrary channels and
calculates scale in the same way as dynamically_quantize_per_tensor
(i.e. with fp64 intermediates) in order to match the cuda reference
function. It also allows different dtypes other than int8.

triton graph of dynamically_quantize_per_channel:
https://gist.github.com/HDCharles/62ecd38aa852a8bc2b658277cf816307

Test Plan: python test_quant_primitives.py TestPerChannelQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757989](https://our.internmc.facebook.com/intern/diff/D45757989)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757989).
